src/bgen/TypeManager.cs

```diff
@@ -413,4 +418,120 @@ public TypeManager (BindingTouch bindingTouch, Assembly apiAssembly, Assembly co
 			NSDirectionalEdgeInsets = ConditionalLookup (platformAssembly, "UIKit", "NSDirectionalEdgeInsets");
 		}
 	}
+
+	public string PrimitiveType (Type t, bool formatted = false)
+	{
+		if (t == System_Void)
+			return "void";
+		if (t == System_Int32)
+			return "int";
+		if (t == System_Int16)
+			return "short";
+		if (t == System_Byte)
+			return "byte";
+		if (t == System_Float)
+			return "float";
+		if (t == System_Boolean)
+			return "bool";
+		if (t == System_Char)
+			return "char";
+		if (t == System_nfloat)
+			return "nfloat";
+
+		return formatted ? FormatType (null, t) : t.Name;
+	}
+
+	public string FormatType (Type usedIn, Type type)
+	{
+		return FormatTypeUsedIn (usedIn is null ? null : usedIn.Namespace, type);
+	}
+
+	public string FormatType (Type usedIn, Type type, bool protocolized)
+	{
+		return FormatTypeUsedIn (usedIn?.Namespace, type, protocolized);
+	}
+
+	public string FormatType (Type usedIn, string @namespace, string name)
+	{
+		string tname;
+		if ((usedIn is not null && @namespace == usedIn.Namespace) || BindingTouch.NamespaceManager.StandardNamespaces.Contains (@namespace))
+			tname = name;
+		else
+			tname = "global::" + @namespace + "." + name;
+
+		return tname;
+	}
+
+	public string FormatTypeUsedIn (string usedInNamespace, Type type, bool protocolized = false)
+	{
+		if (type == System_Void)
+			return "void";
+		if (type == System_SByte)
+			return "sbyte";
+		if (type == System_Int32)
+			return "int";
+		if (type == System_Int16)
+			return "short";
+		if (type == System_Int64)
+			return "long";
+		if (type == System_Byte)
+			return "byte";
+		if (type == System_UInt16)
+			return "ushort";
+		if (type == System_UInt32)
+			return "uint";
+		if (type == System_UInt64)
+			return "ulong";
+		if (type == System_Byte)
+			return "byte";
+		if (type == System_Float)
+			return "float";
+		if (type == System_Double)
+			return "double";
+		if (type == System_Boolean)
+			return "bool";
+		if (type == System_String)
+			return "string";
+		if (type == System_nfloat)
+			return "nfloat";
+		if (type == System_nint)
+			return "nint";
+		if (type == System_nuint)
+			return "nuint";
+		if (type == System_Char)
+			return "char";
+		if (type == System_nfloat)
+			return "nfloat";
+
+		if (type.IsArray)
+			return FormatTypeUsedIn (usedInNamespace, type.GetElementType ()) + "[" + new string (',', type.GetArrayRank () - 1) + "]";
+
+		var interfaceTag = protocolized == true ? "I" : "";
+		string tname;
+		// we are adding the usage of ReflectedType just for those cases in which we have nested enums/classes, this soluction does not
+		// work with nested/nested/nested classes. But we are not writing a general solution because:
+		// 1. We have only encountered nested classes.
+		// 2. We are not going to complicate the code more than needed if we have never ever faced a situation with a crazy complicated nested hierarchy, 
```

Comment:
Fixes: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1907601

```suggestion
		// 2. We are not going to complicate the code more than needed if we have never ever faced a situation with a super complicated nested hierarchy, 
```